### PR TITLE
Add support for circuits and labels in Statevector.equiv

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

Add support for circuits and labels in ``Statevector.equiv``. 

### Details and comments

This missing support can lead to unexpected outcomes such as
```python
qc = QuantumCircuit(1)
Statevector.from_instruction(qc).equiv(qc)  # False
```
Now:
```python
qc = QuantumCircuit(1)
statevec = Statevector.from_instruction(qc)
statevec.equiv(qc)  # True
statevec.equiv('0')  # True
```

The support for `equiv(label)` might not be as necessary as the one for circuits, but since the statevector class also has a `from_label` method it seemed nice to have. It also has a `from_int` method, but that requires two arguments -- which could be passed as a tuple but probably wouldn't be very convenient/readable anymore.

### CODE REVIEW

**Review by chriseclectic — CHANGES_REQUESTED** [review]
Currently `equiv(arg)` is implicitly converted to `equiv(Statevector(arg))` so anything that can be default constructed to a statevector can be used (like for other class methods). I don't think this should support for any other types.

If you want to test from a label or circuit you should explicitly initialize another statevector  like:
```
statevec.equiv(Statevector.from_label('0'))
statevec.equiv(Statevector.from_instruction(qc))
```

For the case of a circuit, if you really want this behaviour we should allow `Statevector(circuit)` to implement the current functionality of `Statevector.from_instruction(circuit)` so it works with all methods of the class, which suggested at some point previously.

With regard specifically to the `equiv` method one could also make the argument that the current implicit conversion is already too generous and anything that isn't already a Statevector object should return `False` like it would for `==`. Then you would have to do `lhs.equiv(Statevector(rhs))` explicitly to check for other object types.

**Comment by Cryoris** [comment]
I like the idea of `Statevector(circuit)`, since it would be equally convenient as (and consistent with) `Operator(circuit)` 👍 

I tend to agree with your last paragraph: the implicit conversions in `equiv` are maybe too generic. They are the reason I though `Statevector(data).equiv(circuit)` would work. So combining these two what do you think of
```python
sv = Statevector(circuit1) 
sv.equiv(circuit2)  # error, not the same type
sv.equiv(Statevector(circuit2))  # works
```

**Comment by chriseclectic** [comment]
@Cryoris I like the idea of combining both like your example

File: qiskit/quantum_info/states/statevector.py [hunk 1]
```diff
@@ -242,10 +247,16 @@ def evolve(self, other, qargs=None):
...
+        """Return True if the statevector of other and self are equivalent up to global phase.
+
+        .. note::
+
+            If ``other`` is a ``QuantumCircuit``, the statevector is constructed by assuming that
+            all qubits in the circuit are initialized to the zero state.
 
         Args:
-            other (Statevector): a statevector object.
+            other (Statevector or vector_like or QuantumCircuit): an object from which a
```
Comment (chriseclectic) [inline qiskit/quantum_info/states/statevector.py:258]: I would just leave the extra types out
```suggestion
            other (Statevector): a statevector or object that can be used to initialize a statevector.
```

File: qiskit/quantum_info/states/statevector.py [hunk 2]
```diff
@@ -36,7 +36,10 @@ def __init__(self, data, dims=None):
         """Initialize a statevector object.
 
         Args:
-            data (vector_like): a complex statevector.
+            data (vector_like or QuantumCircuit): A complex vector, an ``Operator`` with only one
```
Comment (chriseclectic) [inline qiskit/quantum_info/states/statevector.py:39]: You could just explicitly list all the valid types here: np.array, list, Statevector, Operator, QuantumCircuit, Instruction

File: qiskit/quantum_info/states/statevector.py [hunk 3]
```diff
@@ -70,6 +73,8 @@ def __init__(self, data, dims=None):
             if input_dim != 1:
                 raise QiskitError("Input Operator is not a column-vector.")
             self._data = np.ravel(data.data)
+        elif isinstance(data, QuantumCircuit):
```
Comment (chriseclectic) [inline qiskit/quantum_info/states/statevector.py:76]:
```suggestion
        elif isinstance(data, (QuantumCircuit, Instruction)):
```

File: releasenotes/notes/statevector-equiv-ad588688fa260940.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+features:
+  - |
+    Allow constructing :class:`~qiskit.quantum_info.Statevector` objects from circuits, assuming
+    that the qubits are initialized in :math:`|0\rangle`. This also allows to call
+    ``Statevector.equiv`` with a circuit as argument.
```
Comment (chriseclectic) [inline releasenotes/notes/statevector-equiv-ad588688fa260940.yaml:6]: I don't think its necessary to mention the `equiv` part here. Just:
>Allow constructing :class:`~qiskit.quantum_info.Statevector` objects from circuits, assuming
    that the qubits are initialized in :math:`|0\rangle`.

**Review by chriseclectic — CHANGES_REQUESTED** [review]
A couple of suggestions. Could you also update the `DensityMatrix` class in this PR to also initialize from circuits.

File: releasenotes/notes/statevector-equiv-ad588688fa260940.yaml (hunk 4 above)
Comment (Cryoris) [reply releasenotes/notes/statevector-equiv-ad588688fa260940.yaml]: done in 5c59880

File: qiskit/quantum_info/states/statevector.py (hunk 3 above)
Comment (Cryoris) [reply qiskit/quantum_info/states/statevector.py]: done in 5c59880

File: qiskit/quantum_info/states/statevector.py (hunk 2 above)
Comment (Cryoris) [reply qiskit/quantum_info/states/statevector.py]: done in 5c59880

File: qiskit/quantum_info/states/statevector.py (hunk 1 above)
Comment (Cryoris) [reply qiskit/quantum_info/states/statevector.py]: done in 5c59880

File: qiskit/quantum_info/states/statevector.py [hunk 5]
```diff
@@ -242,10 +247,16 @@ def evolve(self, other, qargs=None):
         return Statevector._evolve_operator(ret, other, qargs=qargs)
 
     def equiv(self, other, rtol=None, atol=None):
-        """Return True if statevectors are equivalent up to global phase.
+        """Return True if the statevector of other and self are equivalent up to global phase.
```
[the remark this answers is not in the document]
Comment (Cryoris) [reply qiskit/quantum_info/states/statevector.py]: done in 5c59880

File: qiskit/quantum_info/states/statevector.py [hunk 6]
```diff
@@ -242,10 +247,16 @@ def evolve(self, other, qargs=None):
...
+
+        .. note::
+
+            If ``other`` is a ``QuantumCircuit``, the statevector is constructed by assuming that
```
[the remark this answers is not in the document]
Comment (Cryoris) [reply qiskit/quantum_info/states/statevector.py]: done in 5c59880

**Review by chriseclectic — DISMISSED** [review]

**Comment by chriseclectic** [comment]
@Cryoris Now that this is supported, we could a deprecation warning for `Statevector/DensityMatrix.from_instruction`. But maybe we should wait at least one release with the new version before adding that.

**Review by chriseclectic — APPROVED** [review]
LGTM

**Pull request merged**